README.md
# This is Marvelous Carousel

Marvelous Carousel is a widget that allows you to horizontally or vertically scroll through multiple images or items. It is commonly used in applications where users need to browse a collection of images or rotate through content. For example, you can use Carousel for a product gallery, a news feed, or an event calendar.

![Static Badge](https://img.shields.io/badge/Author-KSPoyraz-blue)
[![Linkedin: Kspoyraz](https://img.shields.io/badge/Kspoyraz-blue?logo=Linkedin&logoColor=fff)][linkedin]
[![Github: Kspo](https://img.shields.io/badge/Kspo-white?logo=Github&logoColor=000)][github]
![GitHub Licence](https://img.shields.io/github/license/kspo/apple_stocks_app_clone?label=Licence)
![GitHub last commit](https://img.shields.io/github/last-commit/kspo/apple_stocks_app_clone?label=Last+Commit)
[![Pub.dev: Kspoyraz](https://img.shields.io/badge/pub.dev_link-marvelous_carousel-magenta?logo=dart&logoColor=magenta)][pub.dev]


## 📸 Screenshots
|Main Screen|
|---|
|<img src="https://raw.githubusercontent.com/kspo/marvelous_carousel/main/screenshots/main.png" style="width:150px" width="150px"/>|
 
 ### Example Screenshots
|Simple|Rotation|Scale Items|
|----|----|----|
|<img src="https://raw.githubusercontent.com/kspo/marvelous_carousel/main/screenshots/simple.gif" style="width:150px" width="150px"/>|<img src="https://raw.githubusercontent.com/kspo/marvelous_carousel/main/screenshots/rotation.gif" style="width:150px" width="150px"/>|<img src="https://raw.githubusercontent.com/kspo/marvelous_carousel/main/screenshots/scale.gif" style="width:150px" width="150px"/>|

|Opacity Animation|Stack|Overscroll|
|----|----|----|
|<img src="https://raw.githubusercontent.com/kspo/marvelous_carousel/main/screenshots/opacity.gif" style="width:150px" width="150px"/>|<img src="https://raw.githubusercontent.com/kspo/marvelous_carousel/main/screenshots/stack.gif" style="width:150px" width="150px"/>|<img src="https://raw.githubusercontent.com/kspo/marvelous_carousel/main/screenshots/overscroll.gif" style="width:150px" width="150px"/>|

## Marvelous Crousel Property List

Here are the details of all attributes:

| Attribute Name   | Description                                                           | Applicable Type |
|:-----------------|:----------------------------------------------------------------------|:----------------|
| pagerType        | Set the MarvelousSlider type, i.e., PagerType.Carousel & PagerType.Stack  | -               |
| reverse          | Reverse the order if set to true                                      | Both            |
| scrollDirection  | Scroll direction of your pager, i.e., Axis.Horizontal & Axis.Vertical | Both            |
| viewPortFraction | Set the view port fraction                                            | Carousel        |
| pageSnapping     | Snap to the selected or current page                                  | Both            |
| physics          | Define ScrollPhysics, affecting scrolling velocity, etc.              | Both            |
| onPageChanged    | Callback that returns the current index whenever an item is selected  | Both            |
| scaleX           | Minimum x-scale of unselected page                                    | Both            |
| scaleY           | Minimum y-scale of unselected page                                    | Both            |
| opacity          | Set opacity of unselected page                                        | Both            |
| rotationX        | Maximum x-rotation of unselected page                                 | Carousel        |
| rotationY        | Maximum y-rotation of unselected page                                 | Carousel        |
| overscroll       | Apply overscroll effect to the last item                              | Stack           |
| key              | Set a key for the pager                                               | Both            |
| children         | List of child widgets to add to the pager                             | -               |
| dotsVisible      | set true if you want to see dots                                      | -               |

## MIT License

Copyright (c) 2023 Kazım Selman Poyraz


[linkedin]: https://www.linkedin.com/in/kaz%C4%B1m-selman-poyraz-0048b7143/
[github]: https://github.com/kspo

[pull_down_button]: https://pub.dev/packages/pull_down_button#pulldownbuttontheme
[marqueer]: https://pub.dev/packages/marqueer
[chart_sparkline]: https://pub.dev/packages/chart_sparkline
[pub.dev]:https://pub.dev/packages/marvelous_carousel

CHANGELOG.md
## [0.0.16] - 2023-09-30.

* Some Bug Fixes

## [0.0.6] - 2023-09-30.

* Fresh new features.

## [0.0.5] - 2023-08-11.

* Dots positions can be set.

## [0.0.4] - 2023-08-11.

* Dots added.

## [0.0.3] - 2023-08-11.

* Minor improvement.

## [0.0.2] - 2023-08-11.

* Minor improvement.

## [0.0.1] - 2023-08-11.

* Minor improvement.


lib/stackslider.dart
library stackslider;

import 'dart:math';

import 'package:flutter/material.dart';

class StackSlider extends StatefulWidget {
  final double viewportFraction;
  final double scaleX;
  final double scaleY;
  final double rotationX;
  final double rotationY;
  final double opacity;
  final List<Widget> children;
  final ValueChanged<int>? onPageChanged;
  final bool pageSnapping;
  final bool reverse;
  final ScrollPhysics physics;
  final Axis scrollDirection;
  final PagerType pagerType;
  final double overscroll;
  final bool dotsVisible;
  final double dotsBottomPx;
  final double margin;

  const StackSlider({
    required this.children,
    this.physics = const BouncingScrollPhysics(),
    this.onPageChanged,
    this.pageSnapping = true,
    this.viewportFraction = 0.85,
    this.scaleX = 1,
    this.scaleY = 1,
    this.opacity = 1,
    this.rotationX = 0,
    this.rotationY = 0,
    this.scrollDirection = Axis.horizontal,
    this.pagerType = PagerType.simple,
    this.dotsVisible = true,
    this.dotsBottomPx = 30.0,
    this.reverse = false,
    this.overscroll = 0,
    this.margin = 0,
    Key? key,
  })  : assert(scaleX <= 1 && scaleX >= 0),
        super(key: key);

  factory StackSlider.symmetric({
    required List<Widget> children,
    ValueChanged<int>? onPageChanged,
    ScrollPhysics physics = const BouncingScrollPhysics(),
    bool pageSnapping = true,
    double viewportFraction = 0.75,
    double scale = 0.75,
    double opacity = 1,
    double rotation = 0,
    Axis direction = Axis.horizontal,
    PagerType pagerType = PagerType.simple,
    bool reverse = false,
    bool dotsVisible = true,
    double dotsBottomPx = 30.0,
    double overscroll = 0,
    double margin = 0,
    Key? key,
  }) {
    return StackSlider(
      pageSnapping: pageSnapping,
      physics: physics,
      viewportFraction: viewportFraction,
      onPageChanged: onPageChanged,
      opacity: opacity,
      scaleX: scale,
      scaleY: scale,
      rotationX: rotation,
      rotationY: rotation,
      scrollDirection: direction,
      pagerType: pagerType,
      reverse: reverse,
      dotsVisible: dotsVisible,
      dotsBottomPx: dotsBottomPx,
      overscroll: overscroll,
      key: key,
      children: children,
    );
  }

  @override
  _StackSliderState createState() => _StackSliderState();
}

enum PagerType { simple, stack }

class _StackSliderState extends State<StackSlider> {
  late PageController _pageController;
  double currentPageValue = 0;
  late Size screenSize;
  Size pagerSize = const Size(0, 0);
  GlobalKey? _pagerKey;

  int activeIndex = 0;

  @override
  void initState() {
    super.initState();
    if (widget.pagerType == PagerType.stack) {
      _pageController = PageController(viewportFraction: 1);
    } else {
      _pageController =
          PageController(viewportFraction: widget.viewportFraction);
    }
    _pageController.addListener(() {
      setState(() {
        currentPageValue = _pageController.page!;
      });
    });
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    screenSize = MediaQuery.of(context).size;
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      child: LayoutBuilder(
        builder: (context, constraints) {
          pagerSize = constraints.biggest;
          return Stack(
            children: [
              PageView.builder(
                key: _pagerKey,
                itemCount: widget.children.length,
                pageSnapping: widget.pageSnapping,
                onPageChanged: (int i) {
                  activeIndex = i;
                  widget.onPageChanged;
                },
                physics: widget.physics,
                controller: _pageController,
                scrollDirection: widget.scrollDirection,
                reverse: widget.reverse,
                itemBuilder: (context, index) {
                  var position = index - currentPageValue;
                  var scaleXValue = max(
                      1 - ((position * (1 - widget.scaleX)).abs()),
                      widget.scaleX);
                  var scaleYValue = max(
                      1 - ((position * (1 - widget.scaleY)).abs()),
                      widget.scaleY);
                  var rotationXValue = (pi / 180) *
                      min(((position * (widget.rotationX)).abs()),
                          widget.rotationX);
                  var rotationYValue = (pi / 180) *
                      min(((position * (widget.rotationY)).abs()),
                          widget.rotationY);
                  var opacityValue = max(
                      1 - ((position * (1 - widget.opacity)).abs()),
                      widget.opacity);
                  if (widget.pagerType == PagerType.simple) {
                    if (position == 0) {
                      return Container(
                          margin:
                              EdgeInsets.symmetric(horizontal: widget.margin),
                          child: widget.children[index]);
                    } else {
                      return Container(
                        margin: EdgeInsets.symmetric(horizontal: widget.margin),
                        child: Opacity(
                          opacity: opacityValue,
                          child: Transform(
                            transform: Matrix4.identity()
                              ..scale(scaleXValue, scaleYValue)
                              ..rotateY(rotationYValue)
                              ..rotateX(rotationXValue),
                            alignment: const Alignment(0, 0),
                            child: widget.children[index],
                          ),
                        ),
                      );
                    }
                  } else if (PagerType.stack == widget.pagerType) {
                    if (index == currentPageValue.floor()) {
                      var opacityValue = max(
                          1 - ((position * (1 - widget.opacity)).abs()),
                          widget.opacity);
                      var scaleXValue = (widget.scaleX +
                          (1 - widget.scaleX) * (1 - (position.abs())));
                      var scaleYValue = (widget.scaleY +
                          (1 - widget.scaleY) * (1 - (position.abs())));
                      var transitionX = widget.scrollDirection ==
                              Axis.horizontal
                          ? (widget.reverse
                              ? ((index - currentPageValue) *
                                  ((pagerSize.width +
                                      ((pagerSize.width * (1 - scaleXValue)) *
                                          2) +
                                      widget.overscroll)))
                              : ((currentPageValue - index) *
                                  ((pagerSize.width -
                                      ((pagerSize.width * (1 - scaleXValue)) *
                                          2) +
                                      widget.overscroll))))
                          : 0.0;
                      var transitionY = widget.scrollDirection == Axis.vertical
                          ? (widget.reverse
                              ? ((index - currentPageValue) *
                                  ((pagerSize.height +
                                      ((pagerSize.height * (1 - scaleXValue)) *
                                          2) +
                                      widget.overscroll)))
                              : ((currentPageValue - index) *
                                  ((pagerSize.height -
                                      ((pagerSize.height * (1 - scaleXValue)) *
                                          2) +
                                      widget.overscroll))))
                          : 0.0;
                      return Container(
                        margin: EdgeInsets.symmetric(horizontal: widget.margin),
                        child: Opacity(
                          opacity: opacityValue,
                          child: Transform(
                            transform: Matrix4.identity()
                              ..scale(scaleXValue ?? 0, scaleYValue ?? 0)
                              ..rotateY(rotationYValue ?? 0)
                              ..rotateX(rotationXValue ?? 0)
                              ..translate(transitionX ?? 0, transitionY ?? 0),
                            alignment: const Alignment(0, 0),
                            child: widget.children[index],
                          ),
                        ),
                      );
                    } else if (index == currentPageValue.floor() + 1) {
                      return Container(
                        margin: EdgeInsets.symmetric(horizontal: widget.margin),
                        child: Transform(
                          transform: Matrix4.identity()..translate(0.0, 0.0),
                          alignment: const Alignment(0, 1),
                          child: widget.children[index],
                        ),
                      );
                    } else {
                      return Container(
                          margin:
                              EdgeInsets.symmetric(horizontal: widget.margin),
                          child: widget.children[index]);
                    }
                  } else {
                    return throw Exception(
                        "pagertype is wrong you set, please set PagerType.simple or PagerType.stack");
                  }
                },
              ),
              widget.dotsVisible
                  ? Positioned(
                      bottom: widget.dotsBottomPx,
                      left: 0,
                      right: 0,
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.center,
                        crossAxisAlignment: CrossAxisAlignment.center,
                        children: widget.children
                            .asMap()
                            .map((index, item) => MapEntry(
                                  index,
                                  AnimatedContainer(
                                    duration: const Duration(milliseconds: 500),
                                    margin: EdgeInsets.only(
                                        right: 10, left: index == 0 ? 10 : 0),
                                    width: activeIndex == index ? 25 : 15,
                                    height: 12,
                                    decoration: BoxDecoration(
                                      color: activeIndex == index
                                          ? Theme.of(context).primaryColor
                                          : Colors.grey,
                                      borderRadius: BorderRadius.circular(50),
                                    ),
                                  ),
                                ))
                            .values
                            .toList(),
                      ),
                    )
                  : const SizedBox(),
            ],
          );
        },
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:marvelous_carousel/marvelous_carousel.dart';

import 'arc_banner.dart';

void main() => runApp(const EntryWidget());

class EntryWidget extends StatelessWidget {
  const EntryWidget({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Finite Coverflow Example',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
        brightness: Brightness.dark,
        splashColor: Colors.transparent,
        highlightColor: Colors.transparent,
      ),
      home: const MyApp(),
    );
  }
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  MyAppState createState() => MyAppState();
}

class MyAppState extends State<MyApp> {
  late final List<Widget> _widgets = [];
  late final List<String> _widgetsTiles = [
    "Simple Carousel",
    "Simple Reverse Carousel",
    "Carousel With Rotation",
    "Scale Items On Change",
    "Opacity Animation On Change",
    "Stack Carousel",
    "Overscroll Stack Carousel",
    "Mixed Animation",
  ];

  late final List<String> _widgetsTilesSubt = [
    "This is just simple carousel. Just attach children and set direction.",
    "Simple Carousel with reverse option. Just Play accordingly.",
    "You can set rotation of carousel items. Visible on page change.",
    "You can set scale of carousel items. Visible on page change.",
    "Opacity Animation of carousel items. It's also visible on change",
    "Stack Carousel animation, Visible on page change.",
    "Overscroll Animation visible on page change",
    "Mixed Animation according to all properties",
  ];

  @override
  void initState() {
    super.initState();
    _widgets.addAll([
      SimpleCarousel(
        isReverse: false,
      ),
      SimpleCarousel(
        isReverse: true,
      ),
      const CarouselWithRotation(),
      const ScaleItemsOnChange(),
      const OpacityAnimationOnChange(),
      const StackCarousel(),
      const OverscrollStackCarousel(),
      const MixedAnimation(),
    ]);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SingleChildScrollView(
        physics: const ClampingScrollPhysics(),
        child: Column(
          children: [
            MovieDetailHeader(testMovie),
            Card(
              margin: const EdgeInsets.only(
                  left: 15, right: 15, top: 30, bottom: 100),
              shape: RoundedRectangleBorder(
                borderRadius: BorderRadius.circular(15),
              ),
              child: ListView.separated(
                separatorBuilder: (context, index) => const Divider(
                  height: 5,
                ),
                shrinkWrap: true,
                itemCount: _widgets.length,
                itemBuilder: (context, index) => ListTile(
                  contentPadding:
                      const EdgeInsets.symmetric(horizontal: 20, vertical: 10),
                  title: Text(
                    _widgetsTiles[index],
                    style: const TextStyle(
                      fontWeight: FontWeight.bold,
                    ),
                  ),
                  trailing: const Icon(Icons.arrow_forward),
                  subtitle: Text(
                    _widgetsTilesSubt[index],
                    style: const TextStyle(height: 1.5),
                  ),
                  horizontalTitleGap: 20,
                  onTap: () {
                    Navigator.push(
                      context,
                      MaterialPageRoute(
                        builder: (context) {
                          return _widgets[index];
                        },
                      ),
                    );
                  },
                ),
                padding: EdgeInsets.zero,
                physics: const NeverScrollableScrollPhysics(),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class MixedAnimation extends StatelessWidget {
  const MixedAnimation({
    super.key,
  });

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      bottomNavigationBar: const SizedBox(
        height: 90,
        child: Card(
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              SizedBox(height: 40),
            ],
          ),
        ),
      ),
      appBar: AppBar(
        title: const Text("Mixed Animation"),
      ),
      body: Center(
        child: SizedBox(
          height: 566,
          child: MarvelousCarousel(
            opacity: 0.2,
            scaleX: 0.5,
            scaleY: 0.5,
            overscroll: 0,
            margin: 10,
            reverse: true,
            pagerType: PagerType.stack,
            scrollDirection: Axis.horizontal,
            children: [1, 2, 3, 4, 5]
                .map(
                  (e) => Card(
                    shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(15),
                    ),
                    child: ClipRRect(
                      borderRadius: BorderRadius.circular(15),
                      child: SizedBox(
                        height: 566,
                        width: 382,
                        child: Image.asset(
                          "images/marvel$e.jpeg",
                          fit: BoxFit.cover,
                        ),
                      ),
                    ),
                  ),
                )
                .toList(),
          ),
        ),
      ),
    );
  }
}

class OverscrollStackCarousel extends StatelessWidget {
  const OverscrollStackCarousel({
    super.key,
  });

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      bottomNavigationBar: const SizedBox(
        height: 90,
        child: Card(
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              SizedBox(height: 40),
            ],
          ),
        ),
      ),
      appBar: AppBar(
        title: const Text("Overscroll Stack Animation"),
      ),
      body: Center(
        child: SizedBox(
          height: 566,
          child: MarvelousCarousel(
            overscroll: -300,
            opacity: 0,
            margin: 20,
            pagerType: PagerType.stack,
            dotsVisible: true,
            scrollDirection: Axis.horizontal,
            children: [1, 2, 3, 4, 5]
                .map(
                  (e) => Card(
                    shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(15),
                    ),
                    child: ClipRRect(
                      borderRadius: BorderRadius.circular(15),
                      child: SizedBox(
                        height: 566,
                        width: 382,
                        child: Image.asset(
                          "images/marvel$e.jpeg",
                          fit: BoxFit.cover,
                        ),
                      ),
                    ),
                  ),
                )
                .toList(),
          ),
        ),
      ),
    );
  }
}

class StackCarousel extends StatefulWidget {
  const StackCarousel({
    super.key,
  });

  @override
  State<StackCarousel> createState() => _StackCarouselState();
}

class _StackCarouselState extends State<StackCarousel> {
  bool _isVertical = true;
  bool _isReverse = false;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      bottomNavigationBar: SizedBox(
        height: 90,
        child: Card(
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              const Text(
                "Vertical",
                style: TextStyle(fontWeight: FontWeight.w700, fontSize: 15),
              ),
              Switch(
                value: _isVertical,
                thumbColor: const MaterialStatePropertyAll<Color>(Colors.black),
                onChanged: (bool value) {
                  setState(() {
                    _isVertical = value;
                  });
                },
              ),
              Container(
                color: Colors.grey,
                width: 1,
                height: 30,
              ),
              const SizedBox(
                width: 10,
              ),
              const Text(
                "Reverse",
                style: TextStyle(fontWeight: FontWeight.w700, fontSize: 15),
              ),
              Switch(
                value: _isReverse,
                thumbColor: const MaterialStatePropertyAll<Color>(Colors.black),
                onChanged: (bool value) {
                  setState(() {
                    _isReverse = value;
                  });
                },
              ),
            ],
          ),
        ),
      ),
      appBar: AppBar(
        title: const Text("Stack Carousel"),
      ),
      body: Center(
        child: SizedBox(
          height: 566,
          child: MarvelousCarousel(
            pagerType: PagerType.stack,
            margin: 15,
            reverse: _isReverse,
            scrollDirection: _isVertical ? Axis.vertical : Axis.horizontal,
            children: [1, 2, 3, 4, 5]
                .map(
                  (e) => Card(
                    shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(15),
                    ),
                    child: ClipRRect(
                      borderRadius: BorderRadius.circular(15),
                      child: SizedBox(
                        height: 566,
                        width: 382,
                        child: Image.asset(
                          "images/marvel$e.jpeg",
                          fit: BoxFit.cover,
                        ),
                      ),
                    ),
                  ),
                )
                .toList(),
          ),
        ),
      ),
    );
  }
}

class OpacityAnimationOnChange extends StatelessWidget {
  const OpacityAnimationOnChange({
    super.key,
  });

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      bottomNavigationBar: const SizedBox(
        height: 90,
        child: Card(
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              SizedBox(height: 40),
            ],
          ),
        ),
      ),
      appBar: AppBar(
        title: const Text("Opacity Animation On Change"),
      ),
      body: Center(
        child: SizedBox(
          height: 566,
          child: MarvelousCarousel(
            opacity: 0.25,
            margin: 10,
            scrollDirection: Axis.horizontal,
            children: [1, 2, 3, 4, 5]
                .map(
                  (e) => Card(
                    shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(15),
                    ),
                    child: ClipRRect(
                      borderRadius: BorderRadius.circular(15),
                      child: SizedBox(
                        height: 566,
                        width: 382,
                        child: Image.asset(
                          "images/marvel$e.jpeg",
                          fit: BoxFit.cover,
                        ),
                      ),
                    ),
                  ),
                )
                .toList(),
          ),
        ),
      ),
    );
  }
}

class ScaleItemsOnChange extends StatelessWidget {
  const ScaleItemsOnChange({
    super.key,
  });

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      bottomNavigationBar: const SizedBox(
        height: 90,
        child: Card(
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              SizedBox(height: 40),
            ],
          ),
        ),
      ),
      appBar: AppBar(
        title: const Text("Scale Items On Change"),
      ),
      body: Center(
        child: SizedBox(
          height: 566,
          child: MarvelousCarousel(
            scaleX: 0.8,
            scaleY: 0.4,
            margin: 10,
            scrollDirection: Axis.horizontal,
            children: [1, 2, 3, 4, 5]
                .map(
                  (e) => Card(
                    shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(15),
                    ),
                    child: ClipRRect(
                      borderRadius: BorderRadius.circular(15),
                      child: SizedBox(
                        height: 566,
                        width: 382,
                        child: Image.asset(
                          "images/marvel$e.jpeg",
                          fit: BoxFit.cover,
                        ),
                      ),
                    ),
                  ),
                )
                .toList(),
          ),
        ),
      ),
    );
  }
}

class CarouselWithRotation extends StatelessWidget {
  const CarouselWithRotation({
    super.key,
  });

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      bottomNavigationBar: const SizedBox(
        height: 90,
        child: Card(
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              SizedBox(height: 40),
            ],
          ),
        ),
      ),
      appBar: AppBar(
        title: const Text("Carousel With Rotation"),
      ),
      body: Center(
        child: SizedBox(
          height: 566,
          child: MarvelousCarousel(
            rotationY: 60,
            margin: 0,
            scrollDirection: Axis.horizontal,
            children: [1, 2, 3, 4, 5]
                .map(
                  (e) => Card(
                    shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(15),
                    ),
                    child: ClipRRect(
                      borderRadius: BorderRadius.circular(15),
                      child: SizedBox(
                        height: 566,
                        width: 382,
                        child: Image.asset(
                          "images/marvel$e.jpeg",
                          fit: BoxFit.cover,
                        ),
                      ),
                    ),
                  ),
                )
                .toList(),
          ),
        ),
      ),
    );
  }
}

class SimpleCarousel extends StatefulWidget {
  bool isReverse;
  SimpleCarousel({
    super.key,
    required this.isReverse,
  });

  @override
  State<SimpleCarousel> createState() => _SimpleCarouselState();
}

class _SimpleCarouselState extends State<SimpleCarousel> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      bottomNavigationBar: SizedBox(
        height: 90,
        child: Card(
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.center,
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              const Text(
                "Reverse",
                style: TextStyle(fontWeight: FontWeight.w700, fontSize: 15),
              ),
              Switch(
                value: widget.isReverse,
                thumbColor: const MaterialStatePropertyAll<Color>(Colors.black),
                onChanged: (bool value) {
                  setState(() {
                    widget.isReverse = value;
                  });
                },
              ),
              const SizedBox(height: 40),
            ],
          ),
        ),
      ),
      appBar: AppBar(
        title: Text("Simple${widget.isReverse ? " Reverse " : " "}Carousel"),
      ),
      body: Center(
        child: SizedBox(
          height: 566,
          child: MarvelousCarousel(
            scrollDirection: Axis.horizontal,
            reverse: widget.isReverse,
            margin: 10,
            children: [1, 2, 3, 4, 5]
                .map(
                  (e) => Card(
                    shape: RoundedRectangleBorder(
                      borderRadius: BorderRadius.circular(15),
                    ),
                    child: ClipRRect(
                      borderRadius: BorderRadius.circular(15),
                      child: SizedBox(
                        height: 566,
                        width: 382,
                        child: Image.asset(
                          "images/marvel$e.jpeg",
                          fit: BoxFit.cover,
                        ),
                      ),
                    ),
                  ),
                )
                .toList(),
          ),
        ),
      ),
    );
  }
}

final Movie testMovie = Movie(
  bannerUrl: 'images/banner.webp',
  posterUrl: 'images/bannerprofile.jpeg',
  title: 'Marvelous Carousel',
  categories: ['Simple', 'Easy', 'Ready to Use'],
);
